src/wp-includes/block-template.php

```diff
@@ -224,10 +223,35 @@ function get_the_block_template_html() {
 		return;
 	}
 
-	$content = $wp_embed->run_shortcode( $_wp_current_template_content );
-	$content = $wp_embed->autoembed( $content );
-	$content = shortcode_unautop( $content );
-	$content = do_shortcode( $content );
+	$content = $_wp_current_template_content;
+
+	/**
+	 * Filters the block template HTML content.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string $content The entire block template HTML content.
+	 */
+	$content = apply_filters( 'the_block_template_html', $content );
```

Comment:
Another option would be to move this filter out from inside `get_the_block_template_html()` to instead apply inside of `template-canvas.php` on the entire HTML page:

```diff
diff --git a/src/wp-includes/template-canvas.php b/src/wp-includes/template-canvas.php
index 55f03267c8..f809b63e36 100644
--- a/src/wp-includes/template-canvas.php
+++ b/src/wp-includes/template-canvas.php
@@ -9,6 +9,20 @@
  * Get the template HTML.
  * This needs to run before <head> so that blocks can add scripts and styles in wp_head().
  */
+
+ob_start(
+	static function( $buffer ) {
+		/**
+		 * Filters the block template HTML content.
+		 *
+		 * @since 6.4.0
+		 *
+		 * @param string $content The entire block template HTML content.
+		 */
+		return apply_filters( 'the_block_template_html', $buffer );
+	}
+);
+
 $template_html = get_the_block_template_html();
 ?><!DOCTYPE html>
 <html <?php language_attributes(); ?>>
```

This would be a step toward implementing Core-43258/Core-58285. Granted, the filter would also need to apply for classic themes as well, and filters being added in this PR for block themes probably shouldn't then also apply to classic themes.

Reply:
Benefits: https://github.com/WordPress/wordpress-develop/pull/5188#discussion_r1324804143

Reply:
Based on the way @felixarntz implemented the filter, trying to move this higher in the execution order to run these after the whole template is generated could cause execution order problems. For example, blocks are being parsed as part of the filter in `get_the_block_template_html()`. It's important that some filters run prior to that parsing, and others run after. When blocks are parsed, they may enqueue scripts and styles to be printed in the head. I don't think we would be able to satisfy each of those requirements by a single filter that runs after the entire template-canvas.

I think adding an additional filter that would allow an output buffer to be implemented for the entire rendered content would be helpful, but doesn't necessarily negate the need to process template content prior to assembling the final HTML document.

Reply:
As for a separate filter for the entire HTML document, I've proposed this as part of the full-page client-side navigation experiment in Gutenberg: https://github.com/WordPress/gutenberg/pull/61212